src/network/prover.rs

```diff
@@ -138,9 +154,194 @@ impl<N: Network, E: Environment> Prover<N, E> {
+    /// All requests must go through this `update`, so that a unified view is preserved.
+    ///
+    pub(super) async fn update(&self, request: ProverRequest<N>) {
+        match request {
+            ProverRequest::PoolRequest(operator_ip, share_difficulty, block_template) => {
+                // Process the pool request message.
+                self.process_pool_request(operator_ip, share_difficulty, block_template).await;
+                // Proceed to register the prover to receive the next block template.
+                self.send_pool_register().await;
+            }
+            ProverRequest::MemoryPoolClear(block) => match block {
+                Some(block) => self.memory_pool.write().await.remove_transactions(block.transactions()),
+                None => *self.memory_pool.write().await = MemoryPool::new(),
+            },
+            ProverRequest::UnconfirmedTransaction(peer_ip, transaction) => {
+                // Ensure the node is not peering.
+                if !self.status.is_peering() {
+                    // Process the unconfirmed transaction.
+                    self.add_unconfirmed_transaction(peer_ip, transaction).await
+                }
+            }
+        }
+    }
+
+    ///
+    /// Sends a `PoolRegister` message to the pool IP address.
+    ///
+    async fn send_pool_register(&self) {
+        if E::NODE_TYPE == NodeType::Prover {
+            if let Some(recipient) = self.address {
+                if let Some(pool_ip) = self.pool {
+                    // Proceed to register the prover to receive a block template.
+                    let request = PeersRequest::MessageSend(pool_ip, Message::PoolRegister(recipient));
+                    if let Err(error) = self.peers_router.send(request).await {
+                        warn!("[PoolRegister] {}", error);
+                    }
+                } else {
+                    error!("Missing pool IP address. Please specify a pool IP address in order to run the prover");
+                }
+            } else {
+                error!("Missing prover address. Please specify an Aleo address in order to prove");
+            }
+        }
+    }
+
+    ///
+    /// Processes a `PoolRequest` message from a pool operator.
+    ///
+    async fn process_pool_request(&self, operator_ip: SocketAddr, share_difficulty: u64, block_template: BlockTemplate<N>) {
+        if E::NODE_TYPE == NodeType::Prover {
+            if let Some(recipient) = self.address {
+                if let Some(pool_ip) = self.pool {
+                    // Refuse work from any pool other than the registered one.
+                    if pool_ip == operator_ip {
+                        // If `terminator` is `false` and the status is not `Peering` or `Mining`
+                        // already, mine the next block.
+                        if !self.terminator.load(Ordering::SeqCst) && !self.status.is_peering() && !self.status.is_mining() {
```

Comment:
nit, probably applicable to other spots as well; we might want to get the `status` once rather than obtain the atomic twice via separate calls to `is_peering` and `is_mining`